sympy/functions/combinatorial/factorials.py

```diff
@@ -173,6 +173,67 @@ def eval(cls, n):
 
                     return Integer(result)
 
+    def _facmod(self, n, q):
+        res, N = 1, int(_sqrt(n))
+
+        # exponent of prime p in n! is e_p(n) = [n/p] + [n/p**2] + ...
+        # for p > sqrt(n), e_p(n) < sqrt(n), the primes with [n/p] = m,
+        # occur consecutively and are grouped together in pw[m] for
+        # simultaneous exponentiation at a later stage
+        pw = [1]*N
+
+        m = 2 # to initialize the if condition below
+        for prime in sieve.primerange(2, n + 1):
+            if m > 1:
+                m, y = 0, n // prime
+                while y:
+                    m += y
+                    y //= prime
+            if m < N:
+                pw[m] = pw[m]*prime % q
+            else:
+                res = res*Mod(Pow(prime, m, evaluate=False), q) % q
+
+        for ex, bs in enumerate(pw):
+            if ex == 0 or bs == 1:
+                continue
+            if bs == 0:
+                return 0
+            res = res*Mod(Pow(bs, ex, evaluate=False), q) % q
+
+        return res
+
+    def _eval_Mod(self, q):
+        x = self.args[0]
+        if x.is_integer and x.is_nonnegative and q.is_integer:
+            aq = abs(q)
+            d = aq - x
+            if d.is_nonpositive:
+                return 0
+            else:
+                isprime = aq.is_prime
+                if d == 1:
+                    '''
+                    Apply Wilson's theorem-if a natural number n > 1
+                    is a prime number, (n-1)! = -1 mod n-and its
+                    inverse-if n > 4 is a composite number,
+                    (n-1)! = 0 mod n
+                    '''
+                    if isprime:
+                        return -1 % q
+                    elif isprime == False and (aq - 6).is_nonnegative:
+                        return 0
+                elif x.is_Integer and q.is_Integer:
+                    if isprime and (d - 1 < x):
+                        fc = self._facmod(d - 1, aq)
+                        fc = Mod(Pow(fc, aq - 2, evaluate=False), aq)
```

Comment:
Will this be automatically evaluated by the three argument `pow`?

Reply:
Yes, that could be used.
Though, `_eval_Mod` in `Pow` was optimized for larger powers (may not be of much relevance here).

Reply:
Is this more efficient than `mod_inverse`?

Reply:
Asymptotically it's the same.
Extended Euclidean or Modular Exponentiation, both run in logarithmic time.

Reply:
Using `pow(a, b, c)` would provide better readability if not better performance.